client/src/templates/Challenges/redux/code-storage-epic.js

```diff
@@ -71,17 +78,26 @@ function clearCodeEpic(action$, state$) {
   );
 }
 
-function saveCodeEpic(action$, state$) {
+function userManualSaveEpic(action$, state$) {
+  return action$.pipe(
+    ofType(types.saveCode),
+    // do not save challenge if code is locked
+    filter(() => {
+      return !isCodeLockedSelector(state$.value);
+    }),
+    tap(() => saveToLocalStorage(state$)),
+    ignoreElements()
+  );
+}
+
+function saveOnExecuteChallengeEpic(action$, state$) {
   return action$.pipe(
     ofType(types.executeChallenge),
```

Comment:
`ofType` accepts multiple action types so `ofType(types.executeChallenge, types.saveCode)` will run the epic on execute and save, reducing the duplication